File: glenn/697/ruby/find_shortest_sub_array.rb
# @param {Integer[]} nums
# @return {Integer}
def find_shortest_sub_array(nums)
  first_index = {}
  degree = {}
  max_degree = -1
  min_span = -1

  nums.each_with_index do |val, idx|
    if first_index[val].nil?
      first_index[val] = idx
      degree[val] = 0
    end
    degree[val] = degree[val] + 1
    span = idx - first_index[val] + 1

    if degree[val] > max_degree || (degree[val] == max_degree && span < min_span)
      max_degree = degree[val]
      min_span = span
    end
  end

  min_span
end

# Runtime: 88 ms, faster than 66.67% of Ruby online submissions for Degree of an Array.
# Memory Usage: 212.6 MB, less than 44.44% of Ruby online submissions for Degree of an Array.

File: glenn/697/go/find_shortest_sub_array.go
package glenn

type summaryEntry struct {
	firstIndex int
	degree int
}

func findShortestSubArray(nums []int) int {

	summary := make(map[int]*summaryEntry)

	//degree := make(map[int]int, 0)
  maxDegree := -1
  minSpan := -1

	for idx, val := range nums {
		if _, ok := summary[val]; !ok {
			summary[val] = &summaryEntry{
				firstIndex: idx,
				degree: 0,
			}
		}
		(summary[val]).degree = summary[val].degree + 1
		span := idx - summary[val].firstIndex + 1

		if summary[val].degree > maxDegree || (summary[val].degree == maxDegree && span < minSpan) {
			maxDegree = summary[val].degree
			minSpan = span
		}
	}

	return minSpan
}

// Runtime: 24 ms, faster than 98.41% of Go online submissions for Degree of an Array.
// Memory Usage: 6.7 MB, less than 14.29% of Go online submissions for Degree of an Array.
// Fast but a tad memory hungry
